packages/@aws-cdk/aws-elasticsearch/lib/domain.ts

```diff
@@ -0,0 +1,1503 @@
+   * are attached to data nodes in the Amazon ES domain.
+   *
+   * @default - 10 GiB General Purpose (SSD) volumes per node.
+   */
+  readonly ebs?: EbsOptions;
+
+  /**
+   * The cluster capacity configuration for the Amazon ES domain.
+   *
+   * @default - 1 r5.large.elasticsearch data node; no dedicated master nodes.
+   */
+  readonly capacity?: CapacityConfig;
+
+  /**
+   * The cluster zone awareness configuration for the Amazon ES domain.
+   *
+   * @default - no zone awareness (1 AZ)
+   */
+  readonly zoneAwareness?: ZoneAwarenessConfig;
+
+  /**
+   * The Elasticsearch version that your domain will leverage.
+   *
+   * Per https://aws.amazon.com/elasticsearch-service/faqs/, Amazon Elasticsearch Service
+   * currently supports Elasticsearch versions 7.7, 7.4, 7.1, 6.8, 6.7, 6.5, 6.4, 6.3, 6.2,
+   * 6.0, 5.6, 5.5, 5.3, 5.1, 2.3, and 1.5.
+   *
+   */
+  readonly version: ElasticsearchVersion;
+
+  /**
+   * Encryption at rest options for the cluster.
+   *
+   * @default - No encryption at rest
+   */
+  readonly encryptionAtRest?: EncryptionAtRestOptions;
+
+
+  /**
+   * Configuration log publishing configuration options.
+   *
+   * @default - No logs are published
+   */
+  readonly logging?: LoggingOptions;
+
+
+  /**
+   * Specify true to enable node to node encryption.
+   * Requires Elasticsearch version 6.0 or later.
+   *
+   * @default - Node to node encryption is not enabled.
+   */
+  readonly nodeToNodeEncryption?: boolean;
+
+  /**
+   * The hour in UTC during which the service takes an automated daily snapshot
+   * of the indices in the Amazon ES domain. Only applies for Elasticsearch
+   * versions below 5.3.
+   *
+   * @default - Hourly automated snapshots not used
+   */
+  readonly automatedSnapshotStartHour?: number;
+
+  /**
+   * The virtual private cloud (VPC) configuration for the Amazon ES domain. For
+   * more information, see [VPC Support for Amazon Elasticsearch Service
+   * Domains](https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-vpc.html)
+   * in the Amazon Elasticsearch Service Developer Guide.
+   *
+   * @default - VPC not used
+   */
+  readonly vpcOptions?: VpcOptions;
+
+  /**
+   * True to require that all traffic to the domain arrive over HTTPS.
+   *
+   * @default - false
+   */
+  readonly enforceHttps?: boolean;
+
+  /**
+   * The minimum TLS version required for traffic to the domain.
+   *
+   * @default - TLSSecurityPolicy.TLS_1_0
+   */
+  readonly tlsSecurityPolicy?: TLSSecurityPolicy;
+
+  /**
+   * Specifies options for fine-grained access control.
+   * Requires Elasticsearch version 6.7 or later. Enabling fine-grained access control
+   * also requires encryption of data at rest and node-to-node encryption, along with
+   * enforced HTTPS.
+   *
+   * @default - fine-grained access control is disabled
+   */
+  readonly fineGrainedAccessControl?: AdvancedSecurityOptions;
```

Comment:
It's optional in cloudformation as well.

ES access is configured via access policies https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-ac.html#es-ac-types-resource. Fine grained access is another layer on top of that. https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/fgac.html

Reply:
So lets say I use all default configurations of `Domain`, basically:

```ts
new es.Domain(this, 'Domain', { version: es.ElasticsearchVersion.V7_7 })
```

How is a user expected to access the cluster now?

Reply:
You need to sigv4 sign the requests using a role that has permissions to access the cluster (like an admin role).

Reply:
Sorry i'm having a real problem grasping this, and I suspect our users will too. 

You're saying the by default, to perform any query, I would need to use the admin role and sigv4 sign the request? What happens if i configure a master user in `fineGrainedAccessPolicy`? Will I be able to use a normal username and password? Or will I need to also change the global access policies somehow? 

P.S we can hop on a quick zoom to hash this out quicker is you're up for it.

Reply:
I can't do the zoom call now. It's getting late for me. I'll reach out to you in my morning and see if you are still around.

I find it confusing as well and am just trying to piece together what the docs say https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/fgac.html.

So if you don't configure any resource policies on the ES cluster then all your requests need to be sigv4 signed otherwise they are rejected.

You can set up a resource policy to allow non sigv4 signed requests. If you do that and enable fine grained auth then you can do a basic auth request to the cluster with the master user and password.

Reply:
Have a look at the two flowcharts here https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/fgac.html#fgac-access-policies. It's a whole bunch of layers of security based on resource policy, VPC and fine grained access.

Reply:
I'll have a look. Just so that we are aligned, I think that if we provide an initial experience where users easily create the cluster, but are now left hanging not knowing how to actually connect to it, that creates frustration. 

Ideally, I would love it if we could make it so the user is able to run requests with a basic username/password, without signing if thats what they wish.

Do you think that is going to be hard to implement? From what you said, it would basically require us to create a default access policy that allows non signed requests, and enable fineGrainedAccessControl with a default username and a generated password. 

To make it a little more secure, we won't make this the default, but instead maybe have a property like 

```ts
new es.Domain(this, 'Domain', {
  version: es.ElasticsearchVersion.V7_7,
  enableBasicDefaultAuth: true
})
``` 

I don't care how ugly the name of it is, since we can change that later on. 

WDYT?

Reply:
I'm happy to create some default resource policies that users can specify.

Personally, I think having every request SIGV4 signed and it tieing in with the AWS policies is a nice feature. I think removing that and just using basic auth with a single user account makes the cluster less secure. I do understand wanting to have a very simple starting point though.

The update behaviour for enabling/disabling fine grained access is replacement https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-elasticsearch-domain-advancedsecurityoptionsinput.html. So once it's enabled it can't be turned off. Although I have no idea what the result is of enabling it and then removing the master user.

Reply:
@stephnr 

> I think removing that and just using basic auth with a single user account makes the cluster less secure

Definitely, thats why we will make this opt in.

> The update behaviour for enabling/disabling fine grained access is replacement

Thats ok, the default cluster we provide is already configured for development, so I believe a cluster replacement is inevitable when a user chooses to move to production.

So lets create this basic setup for the user if he/she so chooses.

Appreciate you and @adamelmore patience and thorough work here!

Reply:
So I started working on this and I have a bit of an issue. Right now we have a bug where we accept access policies but don't actually pass them on.To specify the access policy one needs to reference the domain, which doesn't exist at that stage. I tried to have a look at how some other CDK constructs deal with this and couldn't find ones that take an access policy.

The reason this is tripping me up for this is that I'm trying to do this:

```
  /**
   * Configures the domain so that unsigned basic auth is enabled. If no master user is provided a default master user
   * with username `admin` and a dynamically generated password stored in KMS is created. The password can be retrieved
   * by getting `masterUserPassword` from the domain instance.
   *
   * @param the initial domain configuration
   * @returns the initial domain configuration modified to set up unsigned basic auth.
   */
  public static withUnsignedBasicAuth(props: DomainProps): DomainProps {
    if (parseVersion(props.version) < 6.7) {
      throw new Error('Using unsigned basic auth requires Elasticsearch version 6.7 or later.');
    }

    const masterUserArn = props.fineGrainedAccessControl?.masterUserArn;
    const masterUserName = props.fineGrainedAccessControl?.masterUserName;

    const unsignedAccess = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ['es:ESHttp*'],
      principal: new iam.Anyone(),
      //TODO: resource
    })

    return {
      ...props,
      accessPolicies: props.accessPolicies.concat(unsignedAccess),
      fineGrainedAccessControl: {
        ...props.fineGrainedAccessControl,
        masterUserName: masterUserArn == null ? (masterUserName ?? 'admin') : null,
      },
      enforceHttps: true,
      nodeToNodeEncryption: true,
      encryptionAtRest: {
        ...props.encryptionAtRest,
        enabled: true,
      },
    };
  }
```

I started with trying to add a boolean `enabledUnsignedBasicAuth` to `DomainProps` but I didn't like that. In order to have that boolean value I would need to:

1. Check the status of `encryptionAtRest`. If it is set to false I need to error, if it not set I need to enable it.
1. Same for `nodeToNodeEncryption`.
1. Same for `enforceHttp`.
1. I then need to check the status of the fine grained access control if there are existing configuration I need to respect them. Otherwise set default values.
1. Check that the right version is used.
1. Last I need to add a new access policy that allows unsigned access to the list of access policies.

I am somewhat uncomfortable with one property value changing 5 other property values. I think that is quite messy and I personally would find that to be confusing behaviour.

I prefer having a function that enables this on a domain (I don't think that is possible since it needs to be at construction), a function that enables this at the props level (it's what I'm trying above) or a set of default settings with the right properties set.

Reply:
@stephanh

> To specify the access policy one needs to reference the domain, which doesn't exist at that stage

Im not sure I follow, the JSON in the policy needs to know the address of the domain? Or are you just talking about how to structure the flow like you described above? 

I agree that having a property called `enableBasicAuth` along with the regular domain props might be confusing since it effectively has to set other properties that might be configured to conflicting values. 

In that sense, I feel the solution you proposed is also a bit awkward because the input to the function may also contain these conflicting properties.

How about we have a separate construct called `BasicAuthDomain` that doesn't expose `encryptionAtRest` and all the others...?

Im honestly ok with anything you come up with that works, we can iterate on the API later on.

Reply:
I had a chat with @iliapolo I am going to:

1. Add a property to DomainProps for enabling basic auth.
1. When the property is set to true I will check any of the other required properties. If they are filled in with incompatible values I'll throw an error. If they are not filled in I will set the right values.
1. Document what is happening.

Reply:
Done.